src/shared/fork-journal.c

```diff
@@ -0,0 +1,203 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "build-path.h"
+#include "escape.h"
+#include "event-util.h"
+#include "exit-status.h"
+#include "fd-util.h"
+#include "fork-journal.h"
+#include "notify-recv.h"
+#include "parse-util.h"
+#include "process-util.h"
+#include "signal-util.h"
+#include "socket-util.h"
+#include "strv.h"
+
+static int on_child_exit(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        PidRef *child = ASSERT_PTR(userdata);
+
+        assert(si->si_pid == child->pid);
+
+        log_debug("Child " PID_FMT " died before sending notification message.", child->pid);
+        return -EPROTO;
```

Comment:
sd_event_exit() ?

Reply:
No need, we call sd_event_source_set_exit_on_failure() on the event source, which means we can just return an error here and the right thing happens.